File: python/91_right_triangles_with_integer_coordinates.py
def right_triangles_with_integer_coordinates(max_coord: int) -> int:
    triangles = set()
    for x1 in range(max_coord + 1):
        for y1 in range(max_coord + 1):
            # P(x1, y1) cannot lie at the origin
            if x1 == 0 and y1 == 0:
                continue
            for x2 in range(max_coord + 1):
                for y2 in range(max_coord + 1):
                    # Q(x2, y2) cannot lie at the origin
                    if x2 == 0 and y2 == 0:
                        continue
                    # P and Q must be distinct
                    if x2 == x1 and y2 == y1:
                        continue
                    # avoid double-counting triangles
                    if (x2, y2, x1, y1) in triangles:
                        continue
                    if is_right_triangle(x1, y1, x2, y2):
                        triangles.add((x1, y1, x2, y2))
    
    return len(triangles)
    
def is_right_triangle(x1: int, y1: int, x2: int, y2: int) -> bool:
    # calculate squares of side lengths
    s1: int = x1 * x1 + y1 * y1
    s2: int = x2 * x2 + y2 * y2
    s3: int = (y2 - y1) ** 2 + (x2 - x1) ** 2
    
    # check if an arrangement of sides fulfills a^2 + b^2 = c^2
    return (s1 + s2 == s3) or (s2 + s3 == s1) or (s1 + s3 == s2)

if __name__ == "__main__":
    print(right_triangles_with_integer_coordinates(50))

File: cpp/91_right_triangles_with_integer_coordinates.cpp
#include <iostream>
#include <set>
#include <cmath>

int right_triangles_with_integer_coordinates(){
    std::set<std::pair<std::pair<int, int>, std::pair<int, int>>> triangles;
    for (int px = 0; px <= 50; px++){
        for (int py = 0; py <= 50; py++){
            // point P may lie on axes, but cannot be at origin
            if (px == 0 && py == 0) continue;
            
            for (int qx = 0; qx <= 50; qx++){
                for (int qy = 0; qy <= 50; qy++){
                    // point Q may lie on axes, but cannot be at origin
                    if (qx == 0 && qy == 0) continue;
                    // P and Q must be distinct points
                    if (px == qx && py == qy) continue;
                    // count triangles only once
                    if (triangles.count({{qx, qy}, {px, py}})) continue;
                    
                    // calculate squares of side lengths
                    int a = px * px + py * py;
                    int b = qx * qx + qy * qy;
                    int c = std::pow(qx - px, 2) + std::pow(qy - py, 2);
                    // check if any permutation of squared lengths satisfies Pythagorean theorem
                    if (a + b == c || a + c == b || b + c == a){
                        triangles.insert({{px, py}, {qx, qy}});
                    }
                }
            }
        }
    }
    return triangles.size();
}

int main(){
    std::cout<< right_triangles_with_integer_coordinates() << "\n";
}
